text-serializer-gson/src/main/java/net/kyori/adventure/text/serializer/gson/ComponentSerializerImpl.java

```diff
@@ -70,18 +69,28 @@
   static final String NBT_STORAGE = "storage";
   static final String SEPARATOR = "separator";
 
+  static TypeAdapter<Component> withStyleSerializer(final TypeAdapter<Style> styleSerializer) {
+    return new ComponentSerializerImpl(styleSerializer).nullSafe();
+  }
+
+  private final TypeAdapter<Style> styleSerializer;
+
+  private ComponentSerializerImpl(final TypeAdapter<Style> styleSerializer) {
+    this.styleSerializer = styleSerializer;
+  }
+
   @Override
-  public Component deserialize(final JsonElement json, final Type typeOfT, final JsonDeserializationContext context) throws JsonParseException {
-    return this.deserialize0(json, context);
+  public Component read(final JsonReader in) throws IOException {
+    return this.deserialize(Streams.parse(in));
```

Comment:
> I'd rather you request TypeAdapter instances via a `Gson` instance in each adapter's constructor, rather than using static fields, at least where possible. That allows doing fun tricks with swapping out one adapter for another.

Actually doing this would solve a couple of issues, first this one you pointed out, and second the maze-like mess of the style serializer and component serializer (impl) requiring each other leading to this:
https://github.com/KyoriPowered/adventure/blob/06ec1dfd3cff6daeb72e01b345b43268e47ee074/text-serializer-gson/src/main/java/net/kyori/adventure/text/serializer/gson/StyleSerializer.java#L92-L93

Additionally it would be an ideal/equivalent replacement for `JsonSerializationContext` / `JsonDeserializationContext` give its over-usage in some of the serializers.

That was an eye-opening suggestion :smile:

Reply:
Unfortunately I think you might not be able to avoid the style serializer mess because of the circular dependency there -- but you are welcome to give it a shot :p

Reply:
I believe it should work fine? The circular dependency is not needed at the spot on construction/creation, but rather lazily.
Will definitely try and test! Ending with a SOE from some infinite recursion is not on my schedule :laughing: